spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/DataAccessStrategy.java

```diff
@@ -35,23 +36,59 @@
  * @author Tyler Van Gorder
  * @author Milan Milanov
  * @author Myeonghyeon Lee
+ * @author Chirag Tailor
  */
 public interface DataAccessStrategy extends RelationResolver {
 
 	/**
-	 * Inserts a the data of a single entity. Referenced entities don't get handled.
+	 * Inserts the data of a single entity. Referenced entities don't get handled.
 	 *
+	 * @param <T> the type of the instance.
 	 * @param instance the instance to be stored. Must not be {@code null}.
 	 * @param domainType the type of the instance. Must not be {@code null}.
 	 * @param identifier information about data that needs to be considered for the insert but which is not part of the
 	 *          entity. Namely references back to a parent entity and key/index columns for entities that are stored in a
-	 *          {@link Map} or {@link java.util.List}.
-	 * @param <T> the type of the instance.
+	 *          {@link Map} or {@link List}.
 	 * @return the id generated by the database if any.
 	 * @since 1.1
+	 * @deprecated since 2.4, use {@link #insert(T, Class, Identifier, boolean)}. This will no longer insert as expected
+	 *             when the id property of the instance is pre-populated.
 	 */
 	@Nullable
-	<T> Object insert(T instance, Class<T> domainType, Identifier identifier);
+	@Deprecated
+	default <T> Object insert(T instance, Class<T> domainType, Identifier identifier) {
+		return insert(instance, domainType, identifier, false);
```

Comment:
@schauder I reinstated and deprecated the original method signature for backward compatibility. Is it acceptable that the default implementation of this will no longer work for certain cases or should I try to offer more thorough concrete implementations? If needed, I could reuse the logic from [`WritingContext#idPropertyIsSet`](https://github.com/spring-projects/spring-data-relational/pull/1191/files#diff-574b599c8d79f36deb6620b56a08d08463b998baf9f246b0d1d7fc4e2a6d95b4R307-R320) in `DefaultDataAccessStrategy` to determine a more appropriate value for the `includeId` parameter.